scrapy/core/http2/stream.py

```diff
@@ -0,0 +1,422 @@
+
+        # Clear buffer earlier to avoid keeping data in memory for a long time
+        self._response['body'].truncate(0)
+
+        self.stream_closed_local = True
+        self._conn.reset_stream(self.stream_id, ErrorCodes.REFUSED_STREAM)
+        self.close(reason)
+
+    def _is_data_lost(self) -> bool:
+        assert self.stream_closed_server
+
+        expected_size = self._response['flow_controlled_size']
+        received_body_size = int(self._response['headers'][b'Content-Length'])
+
+        return expected_size != received_body_size
+
+    def close(self, reason: StreamCloseReason, error: Exception = None):
+        """Based on the reason sent we will handle each case.
+
+        Arguments:
+            reason -- One if StreamCloseReason
+        """
+        if self.stream_closed_server:
+            raise StreamClosedError(self.stream_id)
+
+        if not isinstance(reason, StreamCloseReason):
+            raise TypeError(f'Expected StreamCloseReason, received {reason.__class__.__qualname__}')
+
+        self._cb_close(self.stream_id)
+        self.stream_closed_server = True
+
+        flags = None
+        if b'Content-Length' not in self._response['headers']:
+            # Missing Content-Length - {twisted.web.http.PotentialDataLoss}
+            flags = ['partial']
+
+        # NOTE: Order of handling the events is important here
+        # As we immediately cancel the request when maxsize is exceeded while
+        # receiving DATA_FRAME's when we have received the headers (not
+        # having Content-Length)
+        if reason is StreamCloseReason.MAXSIZE_EXCEEDED:
+            expected_size = int(self._response['headers'].get(b'Content-Length', -1))
+            error_msg = (
+                f'Cancelling download of {self._request.url}: expected response '
+                f'size ({expected_size}) larger than download max size ({self._download_maxsize}).'
+            )
+            logger.error(error_msg)
+            self._deferred_response.errback(CancelledError(error_msg))
+
+        elif reason is StreamCloseReason.ENDED:
+            self._fire_response_deferred(flags)
+
+        # Stream was abruptly ended here
+        elif reason is StreamCloseReason.CANCELLED:
+            # Client has cancelled the request. Remove all the data
+            # received and fire the response deferred with no flags set
+
+            # NOTE: The data is already flushed in Stream.reset_stream() called
+            # immediately when the stream needs to be cancelled
+
+            # There maybe no :status in headers, we make
+            # HTTP Status Code: 499 - Client Closed Request
+            self._response['headers'][':status'] = '499'
+            self._fire_response_deferred()
+
+        elif reason in (StreamCloseReason.RESET, StreamCloseReason.CONNECTION_LOST):
+            self._deferred_response.errback(ResponseFailed([
+                error if error else Failure()
```

Comment:
Is this correct? `error` is annotated as `Exception` in the function definition.